lib/PlacesFeed.jsm

```diff
@@ -348,6 +349,32 @@ class PlacesFeed {
     urlBar.addEventListener("paste", checkFirstChange);
   }
 
+  async addBookmark(action) {
+    const {url} = action.data;
+    let data;
+
+    try {
+      // Check if url is an existing bookmark
+      data =  await PlacesUtils.bookmarks.fetch({url});
+    } catch (err) {
+      Cu.reportError(err);
+    }
+    if (data) {
+      // console.log("On Rebookmark", data);
+      this.store.dispatch({type: at.PLACES_LINKS_CHANGED});
+      this.store.dispatch({
+        type: at.PLACES_BOOKMARK_ADDED,
+        data: {
+          bookmarkGuid: data.guid,
+          bookmarkTitle: data.title,
+          url,
+        },
+      });
```

Comment:
@mardak Took a shot at checking first if bookmark exists before calling addBookmark on NewtabUtils. As per places team, this check shouldn't be expensive call. On rebookmark, story shows 'Remove bookmark' state on page refresh, trying to figure why?

Reply:
Just noting from our chat, it doesn't update existing pages as we need to Broadcast the action. However, if we end up in the situation where we incorrectly show "Bookmark" action we'll handle that in https://bugzilla.mozilla.org/show_bug.cgi?id=1531596 so probably good to update that bug title and add your findings.

Reply:
Updated https://bugzilla.mozilla.org/show_bug.cgi?id=1531596#c1